NetTopologySuite.IO.SqlServer/SqlServerSpatialWriter.cs

```diff
@@ -0,0 +1,245 @@
+
+                return value;
+            }
+            set
+            {
+                _emitZ = value.HasFlag(Ordinates.Z);
+                _emitM = value.HasFlag(Ordinates.M);
+            }
+        }
+
+        /// <summary>
+        ///     Gets or sets a value indicating whether to write geography data. If not, geometry data will be written.
+        /// </summary>
+        public virtual bool IsGeography { get; set; }
+
+        /// <summary>
+        ///     Writes a binary representation of a given geometry.
+        /// </summary>
+        /// <param name="geometry"> The geometry </param>
+        /// <returns> The binary representation of geometry </returns>
+        public virtual byte[] Write(IGeometry geometry)
+        {
+            using (var stream = new MemoryStream())
+            {
+                Write(geometry, stream);
+
+                return stream.ToArray();
+            }
+        }
+
+        /// <summary>
+        ///     Writes a binary representation of a given geometry.
+        /// </summary>
+        /// <param name="geometry"> The geometry </param>
+        /// <param name="stream"> The stream to write to. </param>
+        public virtual void Write(IGeometry geometry, Stream stream)
+        {
+            var geography = ToGeography(geometry);
+
+            using (var writer = new BinaryWriter(stream))
+            {
+                geography.WriteTo(writer);
+            }
+        }
+
+        private Geography ToGeography(IGeometry geometry)
+        {
+            if (geometry == null)
+            {
+                return new Geography { SRID = -1 };
+            }
+
+            var geometries = new Queue<(IGeometry, int)>();
+            geometries.Enqueue((geometry, -1));
+
+            // TODO: For geography (ellipsoidal) data, set IsLargerThanAHemisphere
+            var geography = new Geography
+            {
+                SRID = Math.Max(0, geometry.SRID),
+                IsValid = geometry.IsValid
+            };
+
+            while (geometries.Count > 0)
+            {
+                var (currentGeometry, parentOffset) = geometries.Dequeue();
+
+                var figureOffset = geography.Figures.Count;
+                var figureAdded = false;
+
+                switch (currentGeometry)
+                {
+                    case IPoint point:
+                    case ILineString lineString:
+                        figureAdded = addFigure(currentGeometry, FigureAttribute.Line);
+                        break;
+
+                    case IPolygon polygon:
+                        // NB: For geography (ellipsoidal) data, the actual shell is the ring oriented counter-clockwise
+                        figureAdded = addFigure(polygon.Shell, FigureAttribute.Line);
+                        foreach (var hole in polygon.Holes)
+                        {
+                            // NB: For geography (ellipsoidal) data, the actual holes are the rings oriented clockwise
```

Comment:
The behavior here is different than NetTopologySuite.IO.SqlServer2008. There, the counter-clockwise-oriented ring is used as the shell in NTS.

I decided on this implementation instead since it avoids re-arranging polygon rings when you round-trip the data.

This is similar to using WKT on SQL Server which also preserves the order of rings regardless of how they're oriented.

Our guidance is to orient the shell counter-clockwise **and** position it first. This will maximize the interoperability of your data.

Reply:
I’ll do some more testing. I think it will already throw sine the first ring won’t contain the others.

There’s also a case where there is no shell—the full globe with holes in it.

A flag to reposition the shell first would probably be helpful...

Reply:
I see four conflicting requirements:

* I want NTS operations to work (use CCW ring as shell)
* I don't know that the shells have to be CCW (reorient it for me)
* I want to use polygons without a shell that are just holes in the globe (NTS operations won't work, but I know what I'm doing)
* I want to preserve all data (don't reposition rings)

My original approach was to do the same thing as WKT on SQL Server--you need to know how SQL Server works. But I think we should also account for the developers that know NTS but not SQL Server.

I wonder if we should add an enum property `GeographyRingBehavior` to the reader and writer with the following values.

Value | Behavior
Check | Throw when shell isn't CCW, holes aren't CW, or no shell
Fix | Re-position and reorient shell and holes. Throw when no shell
Allow | Treat the first ring as the shell in NTS. Send the rings verbatim to SQL Server. (current behavior)

Check would make a good default since it educates users, but unfortunately, it also requires a lot of extra checking. If it's not the default, we probably don't need it and can get away with just a flag that enables the Fix behavior.

Thoughts?

Reply:
My **immediate** thought is that we should release with at least what you call `Check` (and have it be the default if we implement more than one strategy), since the other strategies only alter how we behave in code paths that would have thrown exceptions otherwise.

`Allow` feels like it may be important to have as well, but only as a means of "taking the training wheels off", so-to-speak, to help cases where the performance of `Check` is inappropriate.

Perhaps the `Check` logic can be public, so that a power user could run it on their data set to identify problems in advance, fix those problems offline, and then feel safe running `Allow` on those validated polygons thereafter.

My immediate reaction to `Fix` is that it can wait.  It seems wiser to have users build "fixes" into the processes that produce this data, rather than relying on doing it at this step.  I haven't thought **too much** about the implications that this would have on the kinds of use cases that EF Core would be most interested in, however.

Reply:
> My **immediate** thought is that we should release with at least what you call `Check` (and have it be the default if we implement more than one strategy), since the other strategies only alter how we behave in code paths that would have thrown exceptions otherwise.

That said, if `Check` is too insane to implement right away, and if EF Core really only needs `Allow`, then it's probably fine to just use `Allow` for 0.x and implement `Check` as a breaking change sometime before we call it 1.x.

Reply:
That sounds good for the writer. I'll add a flag called `SkipGeographyChecks`.

What about the reader? Am I worrying too much about re-positioning rings? The existing IO module just does it; have you ever had complaints? Maybe we should just do it by default (and throw for no shell) to make working with NTS easier, and add a `SkipGeographyFixup` flag to disable it entirely.

Reply:
Updated with these flags.